doc: update interface, --stdin flag, `signtx` (#31005)

cobratbq opened the pull request:
Updates to documentation for External Signer.

- Added mention that `signtransaction` command is no longer primary mechanism.
- Document inter-process communication via `--stdin` flag followed with stdin-content.
- Document `signtx` command followed by Base64-encoded PSBT.

cobratbq committed:
doc: external signer: update interface, --stdin flag, IPC-command signtx, contains updates from #33947

Sjors reviewed (commented):
Good idea to update these docs.

cobratbq commented on doc/external-signer.md:
````diff
@@ -78,6 +78,28 @@ Prerequisite knowledge:
+Usage:
+```
+$ <cmd> --stdin
+[…IPC interaction written to stdin…]
````
Makes sense. Will do.

cobratbq commented on doc/external-signer.md:
````diff
@@ -78,6 +78,28 @@ Prerequisite knowledge:
+[…command and arguments written to stdin…]
+```
+
+#### stdin-command `signtx` (required)
````
Oookaaaaaay .. that changes things considerably. That was not at all clear to me.

Any chance you can drop a link to the place in source-code (assuming it is somewhat concentrated), such that I can have a quick peek. I could search myself, but I suspect you are familiar with the code-base.

Sjors replied:
Most of the action is here: 
- https://github.com/bitcoin/bitcoin/blob/master/src/external_signer.cpp
- https://github.com/bitcoin/bitcoin/blob/master/src/common/run_command.cpp

Sjors commented on doc/external-signer.md:
```diff
@@ -63,29 +65,59 @@ Under the hood this uses a [Partially Signed Bitcoin Transaction](psbt.md).
 ## Signer API
 
-In order to be compatible with Bitcoin Core any signer command should conform to the specification below. This specification is subject to change. Ideally a BIP should propose a standard so that other wallets can also make use of it.
+More recent versions of Bitcoin, around v29 and later, have made significant improvements. It is recommended to use more recent versions.
```
This documentation is shipped with Bitcoin Core. If your change is merged it will be part of the v31 release. So it's pointless to tell the user to use v29 or later.

Sjors replied:
Yes, let's leave it out.

Sjors commented on doc/external-signer.md:
```diff
@@ -63,29 +65,59 @@ Under the hood this uses a [Partially Signed Bitcoin Transaction](psbt.md).
+
+With `<name>` one of `main`, `test`, `signet`, `regtest`.
+
+Previously, flag `--testnet` would be used to indicate "_testnet_" specifically. Presently, all calls of external-signer include `--chain` argument with value.
```
This explanation of previous behaviour seems more appropriate for release notes.

Sjors commented on doc/external-signer.md:
````diff
@@ -96,35 +128,39 @@ A future extension could add an optional return field with device capabilities.
 ```
 
-The command returns a psbt with any signatures.
+With `<psbt>` a Base64-encoded PSBT and updated `<psbt>` result returned.
````
What do you mean by "With `<psbt>`"?

Sjors reviewed (commented):
Looks mostly good, but some text seems off or needlessly verbose.

> I expected that someone would squash upon integrating the work.

Maintainers don't touch the original commits, which is especially nice if you GPG sign them.

Whenever you push an update, just squash the changes right away, reviewers will figure out how to view the diff.

cobratbq commented on doc/external-signer.md:
```diff
@@ -143,36 +179,38 @@ $ <cmd> --fingerprint=00000000 --testnet getdescriptors
 
 The `walletdisplayaddress` RPC reuses some code from `getaddressinfo` on the provided address and obtains the inferred descriptor. It then calls `<cmd> --fingerprint=00000000 displayaddress --desc=<descriptor>`.
 
+For external-signer wallets, spending uses `send` or `sendall`. Bitcoin Core builds a PSBT, calls the signer via stdin with `signtx`, and if signatures are sufficient, finalizes and broadcasts the transaction. If the signer is not connected or cancels, the call fails with an error. For fee-bumping on such wallets, use `psbtbumpfee` when interactive signing is required.
```
I took those from #33947. I'll tweak them a bit.

cobratbq commented on doc/external-signer.md:
```diff
@@ -96,35 +128,39 @@ A future extension could add an optional return field with device capabilities.
-### `signtransaction` (required)
+### `signtx` (required)
+
+`signtx` indicates a PSBT-signing-request, followed by Base64-encoded PSBT-content. Quotes are optional, added in more recent versions. Semantically, there is no difference as Base64-encoded content is an uninterrupted sequence of characters regardless.
```
Dropping.

cobratbq commented on doc/external-signer.md:
````diff
@@ -29,20 +29,22 @@ $ bitcoin-cli enumeratesigners
 
 ```sh
-$ bitcoin-cli createwallet "hww" true true "" true true true
+$ bitcoin-cli -named createwallet wallet_name="hww" disable_private_keys=true external_signer=true
````
Changed. I'm assuming you propose the contemporary way to access the RPC.

Sjors commented:
@cobratbq there's certainly no need to sign commits, but when people do sign, it's nice that we don't break the signature.

Sjors commented on doc/external-signer.md:
```diff
@@ -2,6 +2,8 @@
 
 Bitcoin Core can be launched with `-signer=<cmd>` where `<cmd>` is an external tool which can sign transactions and perform other functions. For example, it can be used to communicate with a hardware wallet.
 
+Interaction with external signer is built upon a specialized construct [Partially Signed Bitcoin Transaction (PSBT)](psbt.md).
```
"is built upon a specialized construct" -> "uses"?

Sjors reviewed (approved):
ACK 15da876e6e7c7445f2700364369ba3d0aa0d959c

Sjors commented:
utACK efc02c50ba1213ffab2541616f437ccfb5fb437d

naiyoma reviewed (commented):
utACK efc02c50ba1213ffab2541616f437ccfb5fb437d

I found this ->( `bitcoin rpc can also be substituted for bitcoin-cli`) to be useful, not sure why it was removed? 

Since the doc is mixing  bitcoin rpc and bitcoin-cli, it might be confusing 

> Best to squash commits when ready to merge.

This can be deleted from the description.

sedited commented:
@cobratbq you have a bunch of unresolved review comments here. Will you be able to get around to them?

Sjors commented:
Would indeed be nice to get this merged soon. I'm working on a prototype Rust implementation of HWI (https://github.com/Sjors/async-hwi/pull/1), so it's useful to have the expected behavior well documented.

cobratbq commented:
All, I'll have a look soon. This has indeed dragged on for a long time. I need to recheck what remaining changes are necessary. I'd love to have it merged.

-------- Original Message --------
On Tuesday, 04/28/26 at 10:46 Sjors Provoost ***@***.***> wrote:

> Sjors left a comment [(bitcoin/bitcoin#33765)](https://github.com/bitcoin/bitcoin/pull/33765#issuecomment-4333703818)
>
> Would indeed be nice to get this merged soon. I'm working on a prototype Rust implementation of HWI ([Sjors/async-hwi#1](https://github.com/Sjors/async-hwi/pull/1)), so it's useful to have the expected behavior well documented.
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/bitcoin/bitcoin/pull/33765#issuecomment-4333703818), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AAOYYVXENU4J56PEAH2F2L34YBVWRAVCNFSM6AAAAACK5XLJWWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHM2DGMZTG4YDGOBRHA).
> Triage notifications on the go with GitHub Mobile for [iOS](https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675) or [Android](https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub).
> You are receiving this because you were mentioned.Message ID: ***@***.***>

cobratbq commented on doc/external-signer.md:
```diff
@@ -142,37 +183,39 @@ $ <cmd> --fingerprint=00000000 --testnet getdescriptors
 
 The `walletdisplayaddress` RPC reuses some code from `getaddressinfo` on the provided address and obtains the inferred descriptor. It then calls `<cmd> --fingerprint=00000000 displayaddress --desc=<descriptor>`.
 
+For external-signer wallets, spending uses `send` or `sendall`. Bitcoin Core builds a PSBT, calls the signer via stdin with `signtx`, and if signatures are sufficient, finalizes and broadcasts the transaction. If the signer is not connected or cancels, the call fails with an error. For fee-bumping on such wallets, use `psbtbumpfee` to involve an external signer.
```
They're "external-signer wallets", i.e. wallets based on external signers. It's a notation thing ... otherwise it could also be read as "external signer-wallets", which isn't the correct meaning.

naiyoma replied:
I think hyphenating here is okay as we are describing the type of wallet.

Sjors replied:
"external-signer wallets" makes sense to me as well.

achow101 replied:
external-signer-wallets would be more correct, but my preference is no hyphenation at all.

Sjors replied:
Our GUI uses the phrase "watch-only wallets", not "watch-only-wallets" or "watch only wallets".

cobratbq commented on doc/external-signer.md:
````diff
@@ -24,68 +26,108 @@ Follow the hardware manufacturers instructions for the initial device setup, as
 
 ```sh
-$ bitcoin-cli createwallet "hww" true true "" true true true
+bitcoin rpc createwallet wallet_name="hww2" disable_private_keys=true passphrase="" descriptors=true external_signer=true
````
Mentioning it puts emphasis on the fact that there is a possibility to specify a passphrase. I think this example is copied from another contributor. I think there is value in being explicit. Not everyone is as well familiar with Bitcoin software internals. WDYT?

Sjors replied:
There's (currently) no point in setting a password when creating a watch-only wallet. I would also leave out the unused arguments (passphrase and descriptors), as it makes it clear which arguments are actually relevant here.

cobratbq commented on doc/external-signer.md:
```diff
@@ -24,68 +26,108 @@ Follow the hardware manufacturers instructions for the initial device setup, as
-`bitcoin rpc` can also be substituted for `bitcoin-cli`.
+One can confirm creation of external wallet via `getwalletinfo`, which will then report `"external_signer": true`.
+
+You can also execute these commands using `bitcoin-qt` Debug Console instead of using `bitcoin-cli`.
```
Good catch. Fixing using almost exactly your suggestion.

cobratbq commented on doc/external-signer.md:
```diff
@@ -96,35 +138,34 @@ A future extension could add an optional return field with device capabilities.
+`signtx` indicates a PSBT-signing-request, followed by Base64-encoded PSBT-content. Quotes are optional.
 
-The command returns a psbt with any signatures.
+This command reads request `signtx <base64-encoded psbt>` from stdin and MUST return a JSON object. On success, it MUST contain `{"psbt": "<base64-encoded psbt>"}` updated to include any new signatures. On failure, it SHOULD contain `{"error": "<message>"}`. Presence of a key `error` with value `null` is not considered an error.
```
Your comment is unclear. The use-case in which `signtx` is used, is called with `--stdin`. It then sends `signtx <base64-encoded-PSBT>`. To my knowledge, all recent versions do this, `--stdin` is always passed. (Even more, I haven't been able to test with `signtx` passing PSBT via commandline.)

achow101 replied:
The `signtx` command by itself does not read from stdin. To read from stdin, th3 `--stdin` flag must be provided. The sentence "This command reads request `signtx <base64-encoded psbt>` from stdin" is incorrect because the command does not do that without the `--stdin` flag. It is unnecessary to document for each command what it does with respect to stdin, that is already be communicated in the `--stdin` option documentation.

The interface here is that there is a command `signtx` which takes an argument of `psbt`. This is either a command line argument, or on stdin if `--stdin` is provided. The documentation should reflect that, not simply state that it must take it on stdin.

cobratbq replied:
I missed or forgot about this comment, but it is corrected now.

cobratbq commented on doc/external-signer.md:
````diff
@@ -96,35 +138,34 @@ A future extension could add an optional return field with device capabilities.
-$ <cmd> --fingerprint=<fingerprint> (--testnet) signtransaction <psbt>
-base64_encode_signed_psbt
-```
+`signtx` indicates a PSBT-signing-request, followed by Base64-encoded PSBT-content. Quotes are optional.
````
Fair enough. Changing.

cobratbq commented:
Some review comments processed. Pushed new version, rebased on current `master`.

cobratbq commented:
@Sjors @achow101 a call-out to speed this up.

naiyoma reviewed (commented):
ACK c25c185

This might be out of scope, so feel free to ignore.

I found this part of the docs a bit contradictory:

> The command MUST return an (empty) array with at least a fingerprint field.

I don’t think this is always the case, since this command can return an empty list without a fingerprint field:

hwi --emulators --chain regtest enumerate
[]

Also:
```

bitcoin-cli -rpcwallet=<wallet> getnewaddress
bitcoin-cli -rpcwallet=<wallet> walletdisplayaddress <address>
```

I think walletname is clearer than wallet.

Sjors commented:
ACK c25c1851b7c0d9544cb0e9aed8d7f8ec004d2b3e (update: modulo stin: https://github.com/bitcoin/bitcoin/pull/33765#discussion_r3200083192)

cobratbq commented:
> [..]
> 
> I found this part of the docs a bit contradictory:
> 
> > The command MUST return an (empty) array with at least a fingerprint field.
> 
> I don’t think this is always the case, since this command can return an empty list without a fingerprint field:
> 
> hwi --emulators --chain regtest enumerate
>
> []
> 
> [..]

Fair point.

> I think walletname is clearer than wallet.

I looked up an arbitrary man-page of bitcoin-cli and they show `<walletname>`, so "walletname" it is.

cobratbq commented:
I processed https://github.com/bitcoin/bitcoin/pull/33765#issuecomment-4445137259, rebased onto current `master` (no conflicts) and force-pushed the PR-branch. It should be good to merge.

Sjors commented:
I think you missed https://github.com/bitcoin/bitcoin/pull/33765#discussion_r3200083192?

```diff
diff --git a/doc/external-signer.md b/doc/external-signer.md
index c66eb92903..d10dd069e9 100644
--- a/doc/external-signer.md
+++ b/doc/external-signer.md
@@ -139,9 +139,20 @@ A future extension could add an optional return field `reachable`, in case `<cmd
 ### `signtx` (required)

-`signtx` indicates a PSBT signing-request, followed by Base64-encoded PSBT. Quotes are optional.
+`signtx` indicates a PSBT signing request and takes a Base64-encoded PSBT
+argument. Quotes are optional.

-This command reads request `signtx <base64-encoded psbt>` from stdin and MUST return a JSON object. On success, it MUST contain `{"psbt": "<base64-encoded psbt>"}` updated to include any new signatures. On failure, it SHOULD contain `{"error": "<message>"}`. Presence of a key `error` with value `null` is not considered an error.
+Usage:

-PSBT SHOULD include BIP32 derivations. The command SHOULD fail if none of the BIP32 derivations match a key owned by the device.
+```sh
+<cmd> signtx <base64-encoded psbt>
+```
+
+The command MUST return a JSON object. On success, it MUST contain
+`{"psbt": "<base64-encoded psbt>"}` updated to include any new signatures. On
+failure, it SHOULD contain `{"error": "<message>"}`. Presence of a key `error`
+with value `null` is not considered an error.
+
+PSBT SHOULD include BIP32 derivations. The command SHOULD fail if none of the
+BIP32 derivations match a key owned by the device.
```

cobratbq commented:
> I think you missed [#33765 (comment)](https://github.com/bitcoin/bitcoin/pull/33765#discussion_r3200083192)?

I did. Fixed. Again ready to be merged. :-P (Also fixed your comment on empty named argument.)

Sjors commented:
ACK 3381855e51c85e6a002278d67c2f4ab6219caf59

naiyoma reviewed (commented):
ACK 3381855e51c85e6a002278d67c2f4ab6219caf59

micronits : feel free to ignore 

If you’ve included someone’s changes, I would suggest adding them as a co-author instead of adding the PR in the commit message.

Since `signtransaction` is no longer required, I don't think it's necessary to have it here 
```

If so, it calls `<cmd> --fingerprint=00000000 signtransaction <psbt>`. It waits for the device to return a (partially) signed psbt, tries to finalize it and broadcasts the transaction.

```

Sjors commented on doc/external-signer.md:
```diff
@@ -24,107 +26,144 @@ Follow the hardware manufacturers instructions for the initial device setup, as
     {
-        "fingerprint": "00000000"
+        "fingerprint": "00000000",
+        "name": "trezor_t"
```
Unlike in RPC, the command needs to return `model` here. Fixed in https://github.com/bitcoin/bitcoin/pull/35424
